doc/mgmt/gnmi/master_arbitration.md

```diff
@@ -0,0 +1,331 @@
+
+gNMI client - a software application that is used to connect to a network element and manage its configuration and operational data. The gNMI client can be written in any language that supports gRPC.
+
+SDN - Software-defined networking (SDN) is a network architecture that uses software-based controllers or application programming interfaces (APIs) to communicate with underlying hardware infrastructure and direct traffic on a network.
+
+EID -  Election ID. A 128-bit integer. The election ID MUST be guaranteed to be monotonically increasing.But how it is done is out of scope of this document and implementation.
+
+Reference: https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-master-arbitration.md
+
+### Overview
+
+For high availability, a system may run multiple replicas of a gNMI client (controller). Among the replicas, only one client should be elected as master and do gNMI operations that mutate any state on the target. However, in the event of a network partition, there can be two or more replicas thinking themselves as master. For read-only RPCs, this is OK. But if they both call the `Set` RPC, the target may be incorrectly configured by the stale master. Therefore, "Master Arbitration" is needed when multiple clients exist.
+
+**Definition**
+
+When Master Arbitration feature is turned on, the gNMI server receives a proto with the election ID (EID) specified in it. The election ID is a 128-bit integer MUST be guaranteed to be monotonically increasing -- a new master's election ID MUST always be larger than its predecessor. The process through which election IDs are allocated to client replicas is out-of-scope of this specification.
+
+The client carries the election ID in all of its `Set` requests. When the target receives any `Set` request, it looks at the election ID carried with the requests and stores the largest ID it sees. An election ID that is equal to or greater than the currently stored one should be accepted as the master. On the other hand, Set requests with no election ID or with a smaller election ID are rejected and return an error.
+
+**Client Role**
+
+Sometimes, there is a need to partition the config tree among multiple clients
+(where each client in turn could be replicated). This is accomplished by
+assigning a *role* to each client. Master Arbitration is performed on a per-role
+basis. There can be one master for each role, i.e. multiple master clients are
+allowed only if they belong to different roles.
+
+The role is identified by its `id`, which is assigned offline in agreement
+across the entire control plane. The target will use the role ID to arbitrate
+such that each role has one and only one master controller instance.
+
+To simplify for use-cases where multiple roles are not needed, the client can
+leave the role message unset in MasterArbitration. This implies default role.
+All `Set` RPCs that uses a default role are arbitrated in the same group.
```

Comment:
Is there a possibility of a split brain on the Client side where more than one client is sending the same role 'id'. If yes, what is the expectation of the GNMI server? Is there any "deny-list" of clients maintained to deny both the clients if they send the same role ?

Reply:
If more than one clients share the same role "id", the one with the largest "EID" among them will be considered as Master. The SET gRPC requests from other clients will be dropped. And, the GET gRPC requests from non-Master clients is not impacted. Therefore, there is not need for any "deny-list" of clients.

Reply:
Can you explain how are we storing the last received EID in the config-DB and how it gets restored upon gNMI server or switch restart?

Reply:
We do not store EID in config-DB. Once gNMI server, or switch restarts, the stored EID is reset to 0. In such case, controller manager should still keep the highest EID for the previous master controller. In such case, as long as MA is enabled for gNMI server, the same master controller should still be picked by switch. Worst senario will be EID rollover and at the meantime, gNMI server restarts. In such case, we still recommend controller manager to handle the EID of the previous master controller. If gNMI stores some large EID value after restart, no controller may reach switch after EID rollover with MA enabled, since in-coming EID probably will be smaller. So, we don't think keep EID on switch side is a good idea.

Reply:
>> as long as MA is enabled for gNMI server, the same master controller should still be picked by switch
Since the gNMI server restart sets the EID to 0, whichever controller first sends the set request will be accepted by the switch irrespective of the higher EID master controller  previously chosen before the gNMI restart. How do we make sure that the previous master controller is chosen after the gNMI restart?

Reply:
> **as long as MA is enabled for gNMI server, the same master controller should still be picked by switch**
> 
> -> Since the gNMI server restart sets the EID to 0, whichever controller first sends the set request will be accepted by the switch irrespective of the higher EID master controller previously chosen before the gNMI restart. How do we make sure that the previous master controller is chosen after the gNMI restart?

Looks like you answered this question already, thanks!